wave_lang/kernel/wave/execution_engine/buffer_utils.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025 The Wave Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "buffer_utils.h"
+#include <Python.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace {
+struct GILState {
+  GILState() : gstate(PyGILState_Ensure()) {}
+  ~GILState() { PyGILState_Release(gstate); }
+  PyGILState_STATE gstate;
+};
+
+struct PyDeleter {
+  void operator()(PyObject *obj) const { Py_DECREF(obj); }
+};
+
+using PyObjectPtr = std::unique_ptr<PyObject, PyDeleter>;
+} // namespace
+
+extern "C" void _mlir_ciface_wave_get_buffer(MemRef1Di8 *ret,
+                                             PyObject *obj_ptr) {
+  GILState gil_state;
+
+  // Get tensor.data_ptr() method and call it
+  PyObjectPtr data_ptr_method(PyObject_GetAttrString(obj_ptr, "data_ptr"));
+  if (!data_ptr_method) {
+    PyErr_Clear();
+    throw std::runtime_error(
+        "wave_get_buffer: Object does not have 'data_ptr' attribute");
```

Comment:
I'm a bit suspicious mixing C and C++ APIs here (the exception will have to go through nanobind I suppose), should we just invoke further `PyErr` methods to trigger the exception?